### DISCUSSION TITLE
Relax Limited Templates for setting up trigger_variables

### DISCUSSION THREAD

**Discussion by r-t-s** [home-assistant/architecture#592] [General]
I think that it would be safe to allow  templates to access the hass state (i.e. not require limtited templates)  during the attachment of triggers when evaluating trigger_variables.

i.e. when setting up triggers:
```
        if self._trigger_variables:
            try:
                variables = self._trigger_variables.async_render(
                    self.hass,
                    variables,
                    limited= self.hass.state != CoreState.running
                )
```

trigger_variables are only rendered at configuration time. (Actually they are rendered when hass transitions to the running state)
At that time all integrations have been configured. 
So currently in the above code imited would always be false.

Since triggers variables are defined to only be initialized at configuration time, there is no expectation that their value will change as the state of the system changes dynamically.

**Comment by r-t-s** [comment]
The motivation for this is as follows:

Enhancements to allow the Target selector to be used to specify Triggers for the automation.

Let’s review the example Blueprint "Motion Activated Light"
It uses an entity selector for the motion sensor and a target selector for the set of lights.

But what if I want all lights in an area to turn on when any motion sensor in an area is triggered?
We can't do that at this time.

Below are the set of changes that I propose to accomplish this:

1) Update for State Trigger and Numeric State trigger to take a template for entity_id
    See [52851](https://github.com/home-assistant/core/pull/52851)

    The example there will show how we can use the entity selector part of the Target selector to specify inputs (plural) 
    for the state trigger ...
    Almost there. But what if I select an area???

2) Add some template functions:
    target_to_entities - Returns a closure of entities based on selections from an input target selector
    target_to_devices - Returns a closure of device ids based on selections from an input target selector
    target_to_areas - Returns a closure of the areas based on selections from an input target selector
    I am sure I could refactor code from the homeassistant.turn_on/off  services for handling targets to implement there.

Note: Under current rules these template functions would all need to be limited templates as they access the hass state.

3) Remove the restriction that rendering trigger_variables be rendered with limited templates.
   Trigger attachment is performed when the hass enters the running state.
   At this time all integrations are configured ... so entities should all be attached to hass.
   Except those that are attached later via a configuration activity or dynamically discovered by the running system.
   I believe (not sure) that device controllers make recently discovered devices known to hass durig configuration, although they may be offline until they establish a connection after the configuration phase. 
   Allow full access to templates (include the new ones above, expand ...) when rendering the trigger variables.
   Trigger variables are defined to be rendered/initialized only once at configuration time ... so there should be no expectation that their value changes via some later dynamic configuration.

With the combination of the above we are almost there.
   If we pick an area ... we may want to filter the set of entities we process. 
   The blueprint developer would still need to filter the results of target_to_entities. But this could be done with additional template trigger_variables.
   
   In the Motion Activated Light template ... this is accomplished by using the service light. turn_on service ... it ignores not light entities in the area.
   A blueprint developer may also want to turn on/off lights and switches by inspecting switches that appear to be controlling a light (maybe by looking at the icon)
   And not turn on/off variable speed fans (that appear to look like lights, by looking at the icon).

**Comment by balloob** [comment]
The reason we don't do this is because people would expect that it will always reflect the latest `hass` state and not be based upon a snapshot in time.

**Reply by r-t-s** [reply]
The definition of trigger_variables is that they are set (rendered) at a snapshot in time ... i.e. when they are configured.
They may of course be used throughout the hass life cycle.

**Comment by frenck** [comment]
I don't think we should do this, as it will cause a lot of unexpected behavior.

**Reply by r-t-s** [reply]
I will consider this closed. 

Blueprints allow the encapsulation of sophisticated automation logic. But are extremely limited on how they can be triggered.
